osgp/platform/osgp-adapter-domain-smartmetering/src/main/java/org/opensmartgridplatform/adapter/domain/smartmetering/application/services/InstallationService.java

```diff
@@ -69,23 +70,48 @@ public InstallationService() {
   public void addMeter(
       final MessageMetadata messageMetadata, final AddSmartMeterRequest addSmartMeterRequest)
       throws FunctionalException {
+    this.storeMeter(messageMetadata, addSmartMeterRequest);
+    this.sendRequestToOsgpCore(messageMetadata, addSmartMeterRequest);
+  }
+
+  private void storeMeter(
+      final MessageMetadata messageMetadata, final AddSmartMeterRequest addSmartMeterRequest)
+      throws FunctionalException {
     final String organisationId = messageMetadata.getOrganisationIdentification();
     final String deviceId = messageMetadata.getDeviceIdentification();
     LOGGER.debug(
-        "addMeter for organisationIdentification: {} for deviceIdentification: {}",
+        "store meter for organisationIdentification: {} for deviceIdentification: {}",
         organisationId,
         deviceId);
     final SmartMeteringDevice smartMeteringDevice = addSmartMeterRequest.getDevice();
 
     final String deviceIdentification = messageMetadata.getDeviceIdentification();
-
-    this.smartMeterService.validateSmartMeterDoesNotExist(deviceIdentification);
+    final Optional<SmartMeter> existingSmartMeter =
+        this.smartMeterService.validateSmartMeterDoesNotExist(
+            deviceIdentification, addSmartMeterRequest.getOverwrite());
 
     final SmartMeter smartMeter = this.smartMeterService.convertSmartMeter(smartMeteringDevice);
+    if (existingSmartMeter.isPresent()) {
+      if (addSmartMeterRequest.getOverwrite()) {
+        smartMeter.setVersion(existingSmartMeter.get().getVersion());
+        smartMeter.setId(existingSmartMeter.get().getId());
+      } else {
+        LOGGER.error(
+            "ERROR: SmartMeter with device identification {} already exists and overwrite is not enabled",
+            messageMetadata.getDeviceIdentification());
+      }
+    }
     this.smartMeterService.storeMeter(organisationId, addSmartMeterRequest, smartMeter);
+  }
+
+  private void sendRequestToOsgpCore(
+      final MessageMetadata messageMetadata, final AddSmartMeterRequest addSmartMeterRequest) {
     final SmartMeteringDeviceDto requestDto =
-        this.mapperFactory.getMapperFacade().map(smartMeteringDevice, SmartMeteringDeviceDto.class);
-    this.osgpCoreRequestMessageSender.send(requestDto, messageMetadata);
+        this.mapperFactory
+            .getMapperFacade()
+            .map(addSmartMeterRequest.getDevice(), SmartMeteringDeviceDto.class);
+    requestDto.setOverwrite(addSmartMeterRequest.getOverwrite());
+    this.osgpCoreRequestMessageSender.send(requestDto, messageMetadata); // en dan??
```

Comment:
_en dan??_
Is there an open end to the logic?

Reply:
sorry, removed comment.